torch/fx/passes/shape_prop.py

```diff
@@ -28,7 +29,50 @@ class TensorMetadata(NamedTuple):
     is_quantized : bool
     qparams: Dict[str, Any]
 
-def _extract_tensor_metadata(result : torch.Tensor, include_contiguity=True) -> TensorMetadata:
+    # Sparse metadata when layout in [ torch.sparse_coo/csr/csc/bsr/bsc ]
+    #   batch_dim + sparse_dim + dense_dim = ndim = len(shape)
+    #   blocksize for bsr/bsc
+    #   idx_dtype denote the types used for compressed indices and positions
+    layout : torch.layout
+    batch_dim : int
+    sparse_dim : int
+    dense_dim : int
+    blocksize : Optional[Tuple[int, int]]
+    idx_dtype : Optional[torch.dtype]
+
+
+def _extract_sparse_tensor_metadata(
+    t: torch.Tensor,
+) -> Tuple[int, int, int, Optional[Tuple[int, int]], Optional[torch.dtype]]:
+    layout = t.layout
+    sparse_dim = t.sparse_dim()
+    dense_dim = t.dense_dim()
+    batch_dim = t.ndim - dense_dim - sparse_dim
+    if layout is torch.sparse_coo:
+        assert batch_dim == 0  # no batch dim
+        idx_dtype = t.indices().dtype
+    elif layout is torch.sparse_csr or layout is torch.sparse_bsr:
+        assert sparse_dim == 2
+        assert t.col_indices().dtype == t.crow_indices().dtype
```

Comment:
These asserts are unnecessary as sparse compressed tensor invariant checks step does not allow constructing sparse tensors with col/crow_indices having different types.

Reply:
Ah, I like to be very pedantic always, so that this triggers when later people relax constraints elsewhere. But happy to remove if that is the style.

Reply:
Makes sense but see also https://github.com/pytorch/pytorch/pull/117907#discussion_r1515213939 . While introducing the support for different index types in col/crow_indices is feasible, it would be a huge change for PyTorch so that catching the required changes here is a least thing to be worry about :)